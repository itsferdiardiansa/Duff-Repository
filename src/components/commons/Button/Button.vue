<template>
  <button :class="customClass" :disabled="disabled">
    <div :class="`${prefixClass}-button--wrapper`">
      <template v-if="icon.length">
        <div :class="`${prefixClass}-button--icon`">
          <font-awesome-icon :icon="icon" ref="iconEl" />
        </div>
      </template>

      <template v-if="hasSlot || label.length">
        <div
          :class="[`${prefixClass}-button--label`, { 'has-icon': icon.length }]"
        >
          <slot>
            {{ label }}
          </slot>
        </div>
      </template>
    </div>
  </button>
</template>
<script>
import { computed, getCurrentInstance, ref } from 'vue'
import defaultProps from './props'

export default {
  name: 'CButton',
  props: defaultProps,
  emits: ['increment'],
  setup(props, { emit, slots }) {
    let root = getCurrentInstance()
    let iconEl = ref()

    const getVariantClass = prefixBtnClass => {
      const { inverse, variant } = props
      let btnClass = `${prefixBtnClass}--${variant}`

      if (inverse) btnClass += '-inverse'

      return btnClass
    }

    const customClass = computed(() => {
      const {
        data: { prefixClass },
      } = root
      const { size, disabled, rounded, bold, pill } = props
      const btnClass = prefixClass.concat('-button')
      let customClass = [btnClass]

      customClass.push(getVariantClass(btnClass))

      if (size) customClass.push(btnClass.concat('--' + size))

      if (bold) customClass.push('text-bold')

      if (pill) customClass.push('pill')
      else if (!rounded) customClass.push('no-rounded')

      if (disabled) customClass.push('disabled')

      return [customClass.join(' ')]
    })

    const handleClick = e => {
      emit('click', e)
    }

    const hasSlot = computed(() => !!slots.default)

    return {
      customClass,
      handleClick,
      hasSlot,
      iconEl,
    }
  },
}
</script>
<style lang="scss">
@import './variant.scss';
@import './size.scss';

.#{$prefixClass}-button {
  @apply h-9 focus:outline-none rounded px-3 transition duration-300 relative;
  @apply hover:bg-gray-300 text-gray-800 overflow-hidden whitespace-nowrap;

  @include variant;

  @include size;

  &.no-rounded {
    @apply rounded-none;
  }

  &.pill {
    @apply rounded-full;
  }

  &.text-bold {
    @apply font-bold;
  }

  &.disabled {
    @apply disabled:opacity-70 pointer-events-none;
  }

  &--wrapper {
    @apply flex-grow-0 flex flex-wrap items-center justify-center;
  }

  &--label {
    @apply pt-0.5;

    &.has-icon {
      @apply ml-2;
    }
  }
}
</style>
