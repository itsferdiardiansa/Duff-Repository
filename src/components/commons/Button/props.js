export default {
  label: {
    type: String,
    default: '',
  },
  inverse: {
    type: Boolean,
    default: false
  },
  variant: {
    type: String,
    default: 'light',
    validator: function(variant) {
      return ~['primary', 'danger', 'warning', 'dark', 'success', 'light'].indexOf(variant)
    }
  },
  size: {
    type: String,
    default: 'base',
    validator: function(variant) {
      return ~['xs', 'sm', 'base', 'lg', 'xl'].indexOf(variant)
    }
  },
  icon: {
    type: Array,
    default: () => []
  },
  bold: {
    type: Boolean,
    default: false,
  },
  disabled: {
    type: Boolean,
    default: false,
  },
  rounded: {
    type: Boolean,
    default: true,
  },
  pill: {
    type: Boolean,
    default: false
  },
}