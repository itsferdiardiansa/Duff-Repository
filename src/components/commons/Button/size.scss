@mixin size {
  &.xs {
    @apply text-xs py-1.5 px-2;
  }

  &.sm {
    @apply text-sm p-2;
  }

  &.base {
    @apply text-base p-3;
  }

  &.lg {
    @apply text-lg p-4;
  }
}