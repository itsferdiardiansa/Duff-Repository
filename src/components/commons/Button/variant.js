export const buttonSize = {
  sm: 'py-1 px-2',
}