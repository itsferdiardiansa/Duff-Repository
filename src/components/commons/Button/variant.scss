@mixin variant {
  &.default {
    @apply bg-gray-200 hover:bg-gray-300 text-gray-800;
  }

  &.primary {
    @apply bg-blue-700 hover:bg-blue-800 text-gray-200;
  }

  &.danger {
    @apply bg-red-700 hover:bg-red-800 text-gray-200;
  }

  &.warning {
    @apply bg-yellow-500 hover:bg-yellow-600;
  }

  &.success {
    @apply bg-green-500 hover:bg-green-600 text-gray-200;
  }

  &.dark {
    @apply bg-gray-700 hover:bg-gray-800 text-gray-200;
  }
}