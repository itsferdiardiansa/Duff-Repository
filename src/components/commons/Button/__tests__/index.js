import { shallowMount } from '@vue/test-utils'
import Button from '@common/Button'

describe('common/Button', () => {
  let Component
  let prefixClass
  let handleClick = jest.fn()

  beforeAll(() => {
    Component = shallowMount(Button)
    prefixClass = `${Component.rootVM.prefixClass}-button`
  })

  it('will render and match snapshot', () => {
    expect(Component.element).toMatchSnapshot()
  })

  it('will have a label correctly', async () => {
    const label = 'This is common button'

    await Component.setProps({ label })

    expect(Component.find(`.${prefixClass}--label`).text()).toMatch(label)

    expect(Component.element).toMatchSnapshot()
  })

  it('will have a variant color correctly', async () => {
    let variants = ['primary', 'danger', 'warning', 'dark', 'success']

    for (let variant of variants) {
      await Component.setProps({ variant })

      expect(Component.attributes().class).toContain(`${prefixClass}--${variant}`)
      
      expect(Component.element).toMatchSnapshot()
    }
  })

  it('will have a inverse variant color correctly', async () => {
    let variants = ['primary', 'danger']

    for (let variant of variants) {
      await Component.setProps({ variant, inverse: true })

      expect(Component.attributes().class).toContain(`${prefixClass}--${variant}-inverse`)
      
      expect(Component.element).toMatchSnapshot()
    }
  })

  it('will have a size correctly', async () => {
    let sizes = ['xs', 'sm', 'base', 'lg', 'xl']

    for (let size of sizes) {
      await Component.setProps({ size })

      expect(Component.attributes().class).toContain(`${prefixClass}--${size}`)
      
      expect(Component.element).toMatchSnapshot()
    }
  })

  it('will have a icon correctly', async () => {
    let icon = ['fa', 'trash']

    await Component.setProps({ icon })

    expect(Component.vm.iconEl.icon).toMatchObject(icon)
      
    expect(Component.element).toMatchSnapshot()
  })

  it('will have a disabled attribute correctly', async () => {
    await Component.setProps({ disabled: true })

    expect(Component.attributes()).toHaveProperty('disabled')
      
    expect(Component.element).toMatchSnapshot()
  })

  it('will not a rounded correctly', async () => {
    await Component.setProps({ rounded: false })

    expect(Component.attributes().class).toContain('no-rounded')
      
    expect(Component.element).toMatchSnapshot()
  })

  it('will have a pill correctly', async () => {
    await Component.setProps({ pill: true })

    expect(Component.attributes().class).toContain('pill')
      
    expect(Component.element).toMatchSnapshot()
  })

  it('will have a text bold correctly', async () => {
    await Component.setProps({ bold: true })

    expect(Component.attributes().class).toContain('text-bold')
      
    expect(Component.element).toMatchSnapshot()
  })

  it('will trigger click event correctly', async () => {
    await Component.vm.$emit('click', handleClick)
  
    expect(Component.emitted('click').length).toEqual(1)

    expect(Component.element).toMatchSnapshot()
  })
})