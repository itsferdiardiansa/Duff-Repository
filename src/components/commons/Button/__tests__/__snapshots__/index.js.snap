// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common/Button will have a disabled attribute correctly 1`] = `
<button
  class="lk-button lk-button--danger-inverse lk-button--xl disabled"
  disabled=""
>
  <div
    class="lk-button--wrapper"
  >
    <div
      class="lk-button--icon"
    >
      <font-awesome-icon-stub />
    </div>
    <div
      class="lk-button--label has-icon"
    >
      
      This is common button
      
    </div>
  </div>
</button>
`;

exports[`common/Button will have a icon correctly 1`] = `
<button
  class="lk-button lk-button--danger-inverse lk-button--xl"
>
  <div
    class="lk-button--wrapper"
  >
    <div
      class="lk-button--icon"
    >
      <font-awesome-icon-stub />
    </div>
    <div
      class="lk-button--label has-icon"
    >
      
      This is common button
      
    </div>
  </div>
</button>
`;

exports[`common/Button will have a inverse variant color correctly 1`] = `
<button
  class="lk-button lk-button--primary-inverse lk-button--base"
>
  <div
    class="lk-button--wrapper"
  >
    <!--v-if-->
    <div
      class="lk-button--label"
    >
      
      This is common button
      
    </div>
  </div>
</button>
`;

exports[`common/Button will have a inverse variant color correctly 2`] = `
<button
  class="lk-button lk-button--danger-inverse lk-button--base"
>
  <div
    class="lk-button--wrapper"
  >
    <!--v-if-->
    <div
      class="lk-button--label"
    >
      
      This is common button
      
    </div>
  </div>
</button>
`;

exports[`common/Button will have a label correctly 1`] = `
<button
  class="lk-button lk-button--light lk-button--base"
>
  <div
    class="lk-button--wrapper"
  >
    <!--v-if-->
    <div
      class="lk-button--label"
    >
      
      This is common button
      
    </div>
  </div>
</button>
`;

exports[`common/Button will have a pill correctly 1`] = `
<button
  class="lk-button lk-button--danger-inverse lk-button--xl pill disabled"
  disabled=""
>
  <div
    class="lk-button--wrapper"
  >
    <div
      class="lk-button--icon"
    >
      <font-awesome-icon-stub />
    </div>
    <div
      class="lk-button--label has-icon"
    >
      
      This is common button
      
    </div>
  </div>
</button>
`;

exports[`common/Button will have a size correctly 1`] = `
<button
  class="lk-button lk-button--danger-inverse lk-button--xs"
>
  <div
    class="lk-button--wrapper"
  >
    <!--v-if-->
    <div
      class="lk-button--label"
    >
      
      This is common button
      
    </div>
  </div>
</button>
`;

exports[`common/Button will have a size correctly 2`] = `
<button
  class="lk-button lk-button--danger-inverse lk-button--sm"
>
  <div
    class="lk-button--wrapper"
  >
    <!--v-if-->
    <div
      class="lk-button--label"
    >
      
      This is common button
      
    </div>
  </div>
</button>
`;

exports[`common/Button will have a size correctly 3`] = `
<button
  class="lk-button lk-button--danger-inverse lk-button--base"
>
  <div
    class="lk-button--wrapper"
  >
    <!--v-if-->
    <div
      class="lk-button--label"
    >
      
      This is common button
      
    </div>
  </div>
</button>
`;

exports[`common/Button will have a size correctly 4`] = `
<button
  class="lk-button lk-button--danger-inverse lk-button--lg"
>
  <div
    class="lk-button--wrapper"
  >
    <!--v-if-->
    <div
      class="lk-button--label"
    >
      
      This is common button
      
    </div>
  </div>
</button>
`;

exports[`common/Button will have a size correctly 5`] = `
<button
  class="lk-button lk-button--danger-inverse lk-button--xl"
>
  <div
    class="lk-button--wrapper"
  >
    <!--v-if-->
    <div
      class="lk-button--label"
    >
      
      This is common button
      
    </div>
  </div>
</button>
`;

exports[`common/Button will have a text bold correctly 1`] = `
<button
  class="lk-button lk-button--danger-inverse lk-button--xl text-bold pill disabled"
  disabled=""
>
  <div
    class="lk-button--wrapper"
  >
    <div
      class="lk-button--icon"
    >
      <font-awesome-icon-stub />
    </div>
    <div
      class="lk-button--label has-icon"
    >
      
      This is common button
      
    </div>
  </div>
</button>
`;

exports[`common/Button will have a variant color correctly 1`] = `
<button
  class="lk-button lk-button--primary lk-button--base"
>
  <div
    class="lk-button--wrapper"
  >
    <!--v-if-->
    <div
      class="lk-button--label"
    >
      
      This is common button
      
    </div>
  </div>
</button>
`;

exports[`common/Button will have a variant color correctly 2`] = `
<button
  class="lk-button lk-button--danger lk-button--base"
>
  <div
    class="lk-button--wrapper"
  >
    <!--v-if-->
    <div
      class="lk-button--label"
    >
      
      This is common button
      
    </div>
  </div>
</button>
`;

exports[`common/Button will have a variant color correctly 3`] = `
<button
  class="lk-button lk-button--warning lk-button--base"
>
  <div
    class="lk-button--wrapper"
  >
    <!--v-if-->
    <div
      class="lk-button--label"
    >
      
      This is common button
      
    </div>
  </div>
</button>
`;

exports[`common/Button will have a variant color correctly 4`] = `
<button
  class="lk-button lk-button--dark lk-button--base"
>
  <div
    class="lk-button--wrapper"
  >
    <!--v-if-->
    <div
      class="lk-button--label"
    >
      
      This is common button
      
    </div>
  </div>
</button>
`;

exports[`common/Button will have a variant color correctly 5`] = `
<button
  class="lk-button lk-button--success lk-button--base"
>
  <div
    class="lk-button--wrapper"
  >
    <!--v-if-->
    <div
      class="lk-button--label"
    >
      
      This is common button
      
    </div>
  </div>
</button>
`;

exports[`common/Button will not a rounded correctly 1`] = `
<button
  class="lk-button lk-button--danger-inverse lk-button--xl no-rounded disabled"
  disabled=""
>
  <div
    class="lk-button--wrapper"
  >
    <div
      class="lk-button--icon"
    >
      <font-awesome-icon-stub />
    </div>
    <div
      class="lk-button--label has-icon"
    >
      
      This is common button
      
    </div>
  </div>
</button>
`;

exports[`common/Button will render and match snapshot 1`] = `
<button
  class="lk-button lk-button--light lk-button--base"
>
  <div
    class="lk-button--wrapper"
  >
    <!--v-if-->
    <!--v-if-->
  </div>
</button>
`;

exports[`common/Button will trigger click event correctly 1`] = `
<button
  class="lk-button lk-button--danger-inverse lk-button--xl text-bold pill disabled"
  disabled=""
>
  <div
    class="lk-button--wrapper"
  >
    <div
      class="lk-button--icon"
    >
      <font-awesome-icon-stub />
    </div>
    <div
      class="lk-button--label has-icon"
    >
      
      This is common button
      
    </div>
  </div>
</button>
`;
